### DISCUSSION TITLE
Sprint 12

### DISCUSSION THREAD

**Discussion by Justus-at-Tazama** [tazama-lf/General-Issues#293] [General]
## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@rtkay123 - Rodney Kanjala

## Apologies
None

### @cshezi - Cebolenkosi Shezi

**Yesterday:**

**Today:**

### @vorsterk - Kyle Vorster

**Yesterday:**

**Today:**

### @Lenbkr - Len Bekker

**Yesterday:**

**Today:**

### @JeanPierreNell - Jean-Pierre Nell

**Yesterday:**

**Today:**

### @rtkay123 - Rodney Kanjala

**Yesterday:**

**Today:**

### @johanfol - Johan Foley

**Yesterday:**

**Today:**

**Comment by Justus-at-Tazama** [comment]
## 2023/10/09

## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@rtkay123 - Rodney Kanjala

## Apologies
None

## Minutes:

### @cshezi - Cebolenkosi Shezi

**Yesterday:** Started working on clearing rooted indexes. Removed from the library. Refactored some of the rule queries to adopt the change. Created the new version, but not yet merged. ProtoFile did not need to be updated. Only TMS should be affected. TMS updated to remove attribute transposition. PR to be created for TMS.

**Today:** Wrap up and deploy. Demo video and prep.

### @vorsterk - Kyle Vorster

**Yesterday:** Performance benchmarking on $10k. 1015 EPS without elastic. With Elastic peaked 480 EPS. $5k with elastic peaked at 241. Without elastic peaked at 407. Craaaaaazy. CRSP and TADProc do not appear to be affected. TMS and Typology Processors show drastic changes between logging on/off. Spun up the environment for the $10k.

**Today:** Chasing 500 EPS on $10k. Then demo prep. Video. Start benchmarking Bun. Regular high-score run, then also Bun up the $-k's.

### @Lenbkr - Len Bekker

**Yesterday:** Elysia PoC implementation.

**Today:** Elysia PoC implementation, then demo video and prep.

### @JeanPierreNell - Jean-Pierre Nell

**Yesterday:** Wrapping up the SCM. Started working on typology thresholds on the typology processor.

**Today:** Demo video and prep. Will include a demo of the SCM. Much excite, many wow.

### @rtkay123 - Rodney Kanjala

**Yesterday:** Investigating Bun memory issues. Test out the different memory settings for the Docker binary. Memory tweaks were inconclusive. Tried a comparison between various rules to see if varying performance may have an impact. Some online reports on memory leaks, but no direct evidence in our platform.

**Today:** New ticket. Demo video, log a Bun ticket.

### @johanfol - Johan Foley

**Yesterday:** Blink-182. Saw some castles.

**Today:** Demo overview and presentations. Costing per performance. Business feasibility study, etc.

## Parking lot

Memory issue on Bun.

**Comment by Justus-at-Tazama** [comment]
## 2023/10/10

## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@rtkay123 - Rodney Kanjala

## Apologies
None

## Minutes:

### @cshezi - Cebolenkosi Shezi

**Yesterday:** Completed the removal of the rooted index fields. Redeployed everything on SB4 to catch up the environment. Updated the batch deploy script to achieve the complete redeployment. End-to-end up and running.

**Today:** #292 - will check in with Len for prior experience.

### @vorsterk - Kyle Vorster

**Yesterday:** Reporting on the performance benchmarking. Demo video. Reviewing QUZE as a testing tool for monitoring.

**Today:** Performance Benchmarking on Bun on the $10k. Continue investigating the tool.

### @Lenbkr - Len Bekker

**Yesterday:** Wrestling with schema changes. Were attempting manual changes; was unable to find tools for the conversion. Reviewed the original schema creation process. Created a functional schema in a more modular way. There are some incompatibilities between the pain/pacs implementations between Swagger and Ellysia.

**Today:** Wrapping up the Ellysia schema. Testing the PoC.

### @JeanPierreNell - Jean-Pierre Nell

**Yesterday:** Broke SB3, but fixed it so no-one noticed. SCM complete. Testing key-value pairs. Demo stuff.

**Today:** Key-value pairs; Typology thresholds.

### @rtkay123 - Rodney Kanjala

**Yesterday:** Documenting memory issue discoveries. One of the pre-release version of Bun was less resource intensive. Some speculation that the drain is due to locally deployed packages. Results inconclusive so far. Close for now.

**Today:** #281 

### @johanfol - Johan Foley

**Yesterday:** Demo prep. Numbers, business case, spreadsheets, costing per performance.

**Today:** Performance journey documentation for showcasing.

## Parking lot

SCM - Justus to walk through with JP and identify gaps for additional future prioritisation and enhancement.

**Comment by Justus-at-Tazama** [comment]
## 2023/10/11

## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@rtkay123 - Rodney Kanjala

## Apologies
None

## Minutes:

### @cshezi - Cebolenkosi Shezi

**Yesterday:** Started fixing the bug on #292. Consulted with Len, but bug unrelated. Consulted with Rodney as well. Looks like a syntax error, but is a logic error instead. Found and fixed the bug in the library and updated the TMS. Ticket completed. Implemented all TMS changes into the Bun branch. Some conflicts were resolved. Bun branch is tracking with Main now.

**Today:** New work. Review/update the workflow for Bun.

### @vorsterk - Kyle Vorster

**Yesterday:** Performance bunfunruns. Changes affected Jenkins pipelines and Docker files. Started testing - Bun consumes way more resources. Two of the VMs were maxing out. CPU maxed out, RAM idles at 40% and spikes to 60/70%. The previous infrastructure mix is no longer useful as a baseline. Only able to achieve just over 500TPS on the $10k (previous high score here was approximately 1100). Working on Qaze ( https://qaze.app/ ) monitoring tool in support of NATS. Will be able to identify memory leaks and lagging services. Reached out to Discord community - tool is still in (closed) Beta. Super friendly and helpful team. New version published last night.

**Today:** Monitoring tool evaluation and Bun-runs.

### @Lenbkr - Len Bekker

**Yesterday:** Assisted the team with Docker changes. Ensured that all indexes are up to date. Using TypeBox schemas for Ellysia PoC. Can even derive pacs/pain messages directly from the schema. This is the coolest thing ever. Warning: TypeBox implementation will result in breaking changes throughout the platform data pipeline. Performance gains appear promising, but will need to be tested via benchmarking. Considering a separate branch to compare results.

**Today:** Depending on parking lot outcomes.

### @JeanPierreNell - Jean-Pierre Nell

**Yesterday:** Continued with typology processor threshold processing update. Interdiction triggers the process, but both the alert and the interdiction sets the review flag to true. Check the requirement. Testing on local Docker e2e.

**Today:** Rip out transaction configuration handling and direct processing in the TADProc to the review flags.

### @rtkay123 - Rodney Kanjala

**Yesterday:** Removing descriptions from the results. 

**Today:** Removing descriptions.

### @johanfol - Johan Foley

**Yesterday:** Performance benchmarking journey.

**Today:** Design for list management. Dynamic scaling preparation.

## Parking lot

Ellysia PoC - 

Threshold implementation requirement confirmation

Descriptions removal

**Comment by Justus-at-Tazama** [comment]
## 2023/10/09

## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@rtkay123 - Rodney Kanjala

## Apologies
None

## Minutes:

### @cshezi - Cebolenkosi Shezi

**Yesterday:** Got Jest and Bun to work in concert. Unit testing is working via Bun. Went back to working on the GitHub flows.

**Today:** Connect to Arango via GitHub workflow. New ticket.

### @vorsterk - Kyle Vorster

**Yesterday:** Performance benchmarking on Bun does not seem to be achieving the results at the moment. The infrastructure mix seems to be totally different and we may be learning all of the same lessons from Node.js all over again. Let's hammer it for the rest of the week, take a checkpoint on Monday. Regardless of the infrastructure mix, one VM always seems to max out, while the rest idle. The one that maxes out all the time is the one that has the most pods and that's the typology processor for now. Considering applying a constraint at deployment to mitigate the over-run.

**Today:** Carry on with performance benchmarking and particularly evening out the consumption of resources. Reach out to the Bun community for some insights. If Bun doesn't show any gains, we'll try to chase the high score on Node.js again.

### @Lenbkr - Len Bekker

**Yesterday:** Comparing Fastify barebones with Ellysia. Something in Ellysia and Bun is running much slower than expected. Bun appears to be folding under load. Tapped into the Arango DB conference - see changes on the licensing.

**Today:** More tests on native Linux to rule out WSL. Assisting Kyle with a gander and some candor.

### @JeanPierreNell - Jean-Pierre Nell

**Yesterday:** Threshold implementations on the typology processor. Typology processor complete. TADProc complete. Runs fine on local. E2E broke in deployment: DataCache mysteriously busted. Found a latent bug introduced earlier - fixed by Shezi. Running through test cases and wrapping up.

**Today:** Hard stop at 10:00. Wrapping up threshold implementations as far as possible.

### @rtkay123 - Rodney Kanjala

**Yesterday:** Removing reason and description text from the result payloads. Also started updating the tests. Pushed a library update on Main that removes the description attribute entirely and set the reason as optional (to accommodate the error messages). Be aware when implementing the latest version of the library.

**Today:** Updating the tests. Typology and TADProc should also be sorted in no time.

### @johanfol - Johan Foley

**Yesterday:**

**Today:**

## Parking lot

**Comment by Justus-at-Tazama** [comment]
## 2023/10/09

## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@rtkay123 - Rodney Kanjala

## Apologies
None

## Minutes:

### @cshezi - Cebolenkosi Shezi

**Yesterday:** Completed ticket to fix GitHub workflows. Created a full-cycle workflow. Started working on updating the repository files with the SPDX header. Found a tool to update all of the text files in an automated way. Will create stuff on gitHub and share with the team team. 

**Today:** Execute the change to update the text files and add the SPDX header. PRs to go to Rodney.

### @vorsterk - Kyle Vorster

**Yesterday:** No performance benchmarking - interacting with the Tricky people from Bun. Hit them with the old "you said Bun is quicker but it's really not". Waiting for feedback from the Bun people before carrying on to save costs in deploying for no reason. Spun up a NATS CLI to check performance - trying to see if can connect QAZE to the NATS on our cluster. Waiting for community feedback. Spun up a local server - will do documentation. tried to deploy Rodney's Docker image for Jupyter to be able to run benchmarking reporting - found an issue with the port number.

**Today:** Looking at resolving the port number and review Qaze, check in on Bun, and get Jupyter up and running.

### @Lenbkr - Len Bekker

**Yesterday:** Set up native Linux and then the trouble started. The multi-verse of rabbit-holes. Everything seems to be mostly working.

**Today:** J-Meter testing to see if the platform is up and running on Linux properly and then test to run Ellysium (Ellysium is dead, long live Bun (or Node)) and see what's what. Working on compiling a list of QOL improvements.

### @JeanPierreNell - Jean-Pierre Nell

**Yesterday:**

**Today:** HAPPY BIRTHDAY!

### @rtkay123 - Rodney Kanjala

**Yesterday:** Yesterday Rodney was updating the rules to exclude descriptions and reasons. Reviewed all the rules, backed up by Shezi for reviews.

**Today:** Move on to the typology processor and then moving on to the TADProc. More deployments and testing coming today.

### @johanfol - Johan Foley

**Yesterday:**

**Today:**

## Parking lot

ArangoDB licensing conditions are changing to BSL 1.1 from their next release and they have effectively dead-ended our upgrade path for Open Source for the next 4 years. We will be evaluating alternatives to ArangoDB over the next two months and replacing Arango DB entirely in 2024Q1.

**Comment by Justus-at-Tazama** [comment]
## 2023/10/16

## Attendees
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@vorsterk - Kyle Vorster

## Apologies
@johanfol - Johan Foley
@cshezi - Cebolenkosi Shezi
@rtkay123 - Rodney Kanjala

## Minutes:

### @cshezi - Cebolenkosi Shezi

**Yesterday:**

**Today:** Platform training.

### @vorsterk - Kyle Vorster

**Yesterday:** Jupyter and Qaze and waiting for feedback from Bun. Suspect that Bun is not going to be able to achieve the results we're looking for in the short term. The Qaze people seem super helpful and responsive. Modified YAML and deployment scripts. Pause Jupyter for now - struggling to get it to work. Check in with Rodney tomorrow.

**Today:** Test Qaze today - will help with the 3000TPS-chase today.

### @Lenbkr - Len Bekker

**Yesterday:** More node and Bun comparisons. Suspect the Bun framework needs to achieve a little more maturity. Foregone Linux profiling in favor of ProxMox.

**Today:** Put a pin in Bun. Start something new.

### @JeanPierreNell - Jean-Pierre Nell

**Yesterday:** Completed the typology threshold work. Will deploy to SB1.

**Today:** Testing e2e and updated configs on SB1. Ins and outs of car repairs.

### @rtkay123 - Rodney Kanjala

**Yesterday:**

**Today:** Platform training

### @johanfol - Johan Foley

**Yesterday:**

**Today:** On leave.

## Parking lot

**Comment by Justus-at-Tazama** [comment]
## 2023/10/17

## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@rtkay123 - Rodney Kanjala

## Apologies
None

## Minutes:

### @cshezi - Cebolenkosi Shezi

**Yesterday:**

**Today:** Product training

### @vorsterk - Kyle Vorster

**Yesterday:** Running performance benchmarking on the high end environment. Started where we last left off. With the new code-base, our performance has been cut to 60%. Large amounts of dropped transactions as well. Resource consumption is fairly middling. Arango was up by 10%. Current indexing appears insufficient. Indexing "in place" have not been properly followed through to all the collections. Version tags would have been useful to rebuild the environment at a specific point in time.

**Today:** Plan of attack - attempt to roll back the platform to the ProtoBuff merge and pick up the benchmarking from there. Spin up the $10k environment to support the trial-and-error rebuild of the earlier software at lower cost. Start with rolling back the rule versions to just before the pacs-002-relative update. Attempt to roll back just the TMS API for a start on the basis that the internal throughput seems to match the ingestion rate.

### @Lenbkr - Len Bekker

**Yesterday:** Started working on base fixtures for the testing - attempting to find a smart and automated way to add this to the platform for re-use and accessibility. Automated Jenkins and performance testing in isolation. JMeter test to include NATS and Protobuff - opportunity to create a ticket to investigate.

**Today:** Help Kyle.

### @JeanPierreNell - Jean-Pierre Nell

**Yesterday:** Fixed car tyre and rim. Needs to check the axle and wheel realignment today. Car drives fine, but like a crab. Fixed the bug in the TADProc that prevented alerts. Did some work on the GoLang typology processor. Working on the Redactio. Some issues on the typology processor weighting calculation. Investigating.

**Today:** PRs for the threshold change going up for review and then deployment to SB1.

### @rtkay123 - Rodney Kanjala

**Yesterday:**

**Today:** Product training

### @johanfol - Johan Foley

**Yesterday:**

**Today:** Surfing the canals.

## Parking lot

**Comment by Justus-at-Tazama** [comment]
## 2023/10/09

## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@rtkay123 - Rodney Kanjala

## Apologies
None

## Minutes:

### @cshezi - Cebolenkosi Shezi

**Yesterday:**

**Today:** Product training

### @vorsterk - Kyle Vorster

**Yesterday:** Tested throughput on the TMS API on main branch. Able to achieve 2400 TPS - ruled out as the root of the current performance drop. Only able to achieve 1100. There's definitely something in the main branch that is deteriorating performance. Supported by Len and Jason. Len rebuilt a version of the platform to roll back to an earlier and stable branch. Will maintain a library of snapshots for future use again.

**Today:** Performance benchmarking. Keep on rolling out additional functionality until we can identify the culprit.

### @Lenbkr - Len Bekker

**Yesterday:** Helped Kyle. We started looking at the TMS API first and isolated the TMS API/data preparation service to make sure it wasn't a problem. Removed rule 024/026 from the network map - very little impact. Lots of code changes to unwind, but not a lot of features were implemented overall. All branches from the 18th gave an unstable end-to-end, but found a branch with a blend of ProtoBuff and Fastify that was working. Pain pain pain. Kyle/Jason experimenting with the infrastructure mix. Encroaching on the previous performance. Resource consumption has increased beyond the previous baseline.

**Today:** Support Kyle.

### @JeanPierreNell - Jean-Pierre Nell

**Yesterday:** Fun with unit tests on the typology processor and the TADProc. The library includes a breaking change for processing downstream from the typology processing. All tests done - will be deployed on SB1.

**Today:** Building and deploying on SB1. New ticket today!

### @rtkay123 - Rodney Kanjala

**Yesterday:**

**Today:** Product training

### @johanfol - Johan Foley

**Yesterday:** Surfing the canals.

**Today:** Catching up. Check in on performance benchmarking. Admin and planning.

## Parking lot

Network map alignment with NATS.

**Comment by Justus-at-Tazama** [comment]
## 2023/10/19

## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@rtkay123 - Rodney Kanjala

## Apologies
None

## Minutes:

### @cshezi - Cebolenkosi Shezi

**Yesterday:**

**Today:** Product training

### @vorsterk - Kyle Vorster

**Yesterday:** Roll-back platform is working on local with a single typology. Deploying to production with all the rules E2E doesn't work. Logging is enabled, but no errors are being reported.

**Today:** Suggest to test individual rules via Postman and the NATS REST proxy to see if the rules are completing. Keep on troubleshooting the end-to-end with Len. Call in air support as required.

### @Lenbkr - Len Bekker

**Yesterday:** Helping Kyle. Losing hair. Developing mange. Struggling to rebuild the exact build that we were using at our previous high score. Investigated TADProc/CADProc, but confirmed this is not the problem. It seems that a rule might not be completing. Over-the-shoulder process interacting with Kubernetes is time-consuming and inefficient. Checked in on Bun 1.06. Bun/Ellysia is at 95% of Node/Fastify. Bun/Hono was at 103%! (Linux native prox-mox). WSL results are not worth mentioning.

**Today:** Finding pictures of Spider-Man. Assist Kyle. Watching Spawn in the background for inspiration.

### @JeanPierreNell - Jean-Pierre Nell

**Yesterday:** Review requirements for NATS coordination with the network map. Started deploying the platform. Demoed the SCM to JoPacc. Picked up a ticket to update the typology configuration metadata.

**Today:** Close some gaps in the SCM. Internal demo. Test an updated config to see if any code changes are required for the typology configuration metadata.

### @rtkay123 - Rodney Kanjala

**Yesterday:**

**Today:** Product training

### @johanfol - Johan Foley

**Yesterday:**

**Today:** 

## Parking lot

Investigate a method for rebuilding a specific version of the platform at a point in time.

Investigate automated performance testing on Jenkins.

**Comment by JeanPierreNell** [comment]
## 2023/10/09

## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@rtkay123 - Rodney Kanjala

## Apologies
None

## Minutes:

### @cshezi - Cebolenkosi Shezi

**Yesterday:** Platform Training, Having the time of his life.

**Today:** Shadow JP / Look at SCM

### @vorsterk - Kyle Vorster

**Yesterday:** Getting 400 TPS increase when not using time relevant rules. Max 1600 FTPS yesterday. Worked with Len Mostly.

**Today:** Switching back to main and debug what is slow. Add back APM and Debugging.

### @Lenbkr - Len Bekker

**Yesterday:** Len and Kyle investigated broken e2e on Dev. Deployed old versions of the rules. Helped JP with Mystery Timeframs issues. Weird error message. Investigated getting full stack docker to work with private rules. Stuck at the custom rule installation issue. e2e was working again. New Time Relevant rules are problematic on the old code. 

**Today:** Available if Kyle needs it. Investigate build tools for private rules. Looking at self destructing builder pods.

### @JeanPierreNell - Jean-Pierre Nell

**Yesterday:** Looked at missing Timeframs behaviour, worked on testing renamed Typologies - works fine with small networkmap, Justus's networkmap was borked.

**Today:** Fix SCM , Close Gaps + Support Justus's Testing

### @rtkay123 - Rodney Kanjala

**Yesterday:** Platform Training , Rodney is having withdrawal symptoms 

**Today:** Platform Training Review.

### @johanfol - Johan Foley

**Yesterday:**

**Today:**  Spend time with Kyle. Debugging Performance Lack

## Parking lot
